### DISCUSSION TITLE
Logic and operations for recurring events

### DISCUSSION THREAD

**Discussion by intelagense** [Together-100Devs/Together#108] [General]
# How to get the appropriate data into the current calendar view quickly and without downloading and processing the entire database

What do you think about using MongoDB methods against `startDate` and `endDate` key values to request the data for the current calendar view? Then as the user scrolls through the months the additional data will be requested and added to a local object to update the calendar view (+- a few months in both directions for speed.) 

I am roughly estimating that we can fit about 100-150 events with descriptions in under 100Kb of request data with the current schema setup. 

We can use Date libraries to render the data to the views. (I am thinking this will be done client-side from our previous discussions.)

Not yet sure how to implement this yet but here is some pseudocode to start from. 

1. User visits the site and will see the current month.
2. Mongo finds all events that haven't ended yet going back three months (wrap your head around that one 🥴 )
```
let threeMonthsAgo = New Date - 3 months  //today minus 3 months

db.collection.find({
    endDate: {
        $gt: ISODate(threeMonthsAgo), //returns any ongoing events that havent ended yet or in the previous 3 months
    }
    // can also add a startDate less than as well in case the user scrolls back several months or years.
})
```
3. Loop through the found objects
4. Use the startDate and endDate with date-fns helper functions to determine when and where to place dates on calendar. 

Step 4 is actually the hardest and we may not need steps 1-3 for the MVP. 

Let me know if this all sounds reasonable so far. 🙏

**Comment by Caleb-Cohen** [comment]
What if we just set the end date to automatically be 3 months after the start date? Then just render everything with an active date?

**Comment by intelagense** [comment]
Events that are recurring will be embedded as an array of objects in each containing document in the database. I will update this in the whiteboard discussion #9 and it is linked to issue #126
